src/resolve-dependency.ts

```diff
@@ -1,39 +1,46 @@
-const { isAbsolute, resolve, sep } = require('path');
+import { isAbsolute, resolve, sep } from 'path';
+import { Job } from './node-file-trace';
 
 // node resolver
 // custom implementation to emit only needed package.json files for resolver
 // (package.json files are emitted as they are hit)
-module.exports = function resolveDependency (specifier, parent, job, cjsResolve = true) {
-  let resolved;
+export default function resolveDependency (specifier: string, parent: string, job: Job, cjsResolve = true) {
+  let resolved: string | string[] | undefined;
   if (isAbsolute(specifier) || specifier === '.' || specifier === '..' || specifier.startsWith('./') || specifier.startsWith('../')) {
     const trailingSlash = specifier.endsWith('/');
     resolved = resolvePath(resolve(parent, '..', specifier) + (trailingSlash ? '/' : ''), parent, job);
-  }
-  else {
+  } else {
     resolved = resolvePackage(specifier, parent, job, cjsResolve);
   }
-  if (typeof resolved === 'string' && resolved.startsWith('node:')) return resolved;
-  if (Array.isArray(resolved))
+  if (typeof resolved === 'string' && !resolved.startsWith('node:')) {
+    return job.realpath(resolved, parent);
+  } else if (Array.isArray(resolved)) {
     return resolved.map(resolved => job.realpath(resolved, parent));
-  return job.realpath(resolved, parent);
+  }
+  return resolved;
 };
 
-function resolvePath (path, parent, job) {
-  return resolveFile(path, parent, job) || resolveDir(path, parent, job) || notFound(path, parent);
+function resolvePath (path: string, parent: string, job: Job): string | undefined {
+  const result = resolveFile(path, parent, job) || resolveDir(path, parent, job);
+  if (!result) {
+    throw new NotFoundError(path, parent);
+  }
+  return result;
 }
 
-function resolveFile (path, parent, job) {
-  if (path.endsWith('/')) return;
+function resolveFile (path: string, parent: string, job: Job): string | undefined {
+  if (path.endsWith('/')) return undefined;
   path = job.realpath(path, parent);
   if (job.isFile(path)) return path;
   if (job.ts && path.startsWith(job.base) && path.substr(job.base.length).indexOf(sep + 'node_modules' + sep) === -1 && job.isFile(path + '.ts')) return path + '.ts';
   if (job.ts && path.startsWith(job.base) && path.substr(job.base.length).indexOf(sep + 'node_modules' + sep) === -1 && job.isFile(path + '.tsx')) return path + '.tsx';
   if (job.isFile(path + '.js')) return path + '.js';
   if (job.isFile(path + '.json')) return path + '.json';
   if (job.isFile(path + '.node')) return path + '.node';
+  return undefined;
```

Comment:
Should `resolveFile()` return undefined or should it throw NotFoundError?

Reply:
It should return undefined.